src/main/java/org/rascalmpl/maven/AbstractRascalMojo.java

```diff
@@ -0,0 +1,380 @@
+	}
+
+	protected List<Path> locations(List<String> paths) {
+		return paths.stream()
+			.map(Path::of)
+			.collect(Collectors.toCollection(ArrayList::new));
+	}
+
+	@Override
+	public void execute() throws MojoExecutionException {
+		try {
+			if (System.getProperty("rascal." + skipTag + ".skip") != null) {
+				getLog().info("Skipping " + getClass().getName() + " completely");
+				return;
+			}
+
+			Path binLoc = Path.of(bin);
+			Path generatedSourcesLoc = Path.of(generatedSources);
+			List<Path> srcLocs = locations(srcs);
+			List<Path> ignoredLocs = locations(srcIgnores);
+			List<Path> libLocs = locations(libs);
+
+			getLog().info("configuring paths");
+			for (Path src : srcLocs) {
+				getLog().info("\tregistered source location: " + src);
+			}
+
+			for (Path ignore : ignoredLocs) {
+				getLog().warn("\tignoring sources in: " + ignore);
+			}
+
+			getLog().info("Checking if any files need compilation...");
+
+			List<Path> todoList = getTodoList(binLoc, srcLocs, ignoredLocs);
+
+			if (!todoList.isEmpty()) {
+				getLog().info("Stale source files have been found:");
+				for (Path todo : todoList) {
+					getLog().info("\t" + todo);
+				}
+			}
+			else {
+				getLog().info("No stale source files have been found, skipping compilation.");
+				return;
+			}
+
+			libLocs.addAll(collectDependentArtifactLibraries(project));
+
+			for (Path lib : libLocs) {
+				getLog().info("\tregistered library location: " + lib);
+			}
+
+			getLog().info("Paths have been configured.");
+
+			runMain(verbose, todoList, srcLocs, libLocs, generatedSourcesLoc, binLoc).waitFor();
+
+			return;
+		}
+		catch (InterruptedException e) {
+			throw new MojoExecutionException("nested " + mainClass + " was killed", e);
+		}
+		catch (Throwable e) {
+			throw new MojoExecutionException("error launching " + mainClass, e);
+		}
+	}
+
+	/**
+	 * Finds the rascal.jar file that the pom.xml depends on.
+	 * When the current project is rascal itself we resolve to a declared bootstrap
+	 * dependency.
+	 * @throws MojoExecutionException
+	 */
+	protected Path detectedDependentRascalArtifact(Log log, MavenProject project) {
+		try {
+			if (project.getArtifactId().equals("org.rascalmpl:rascal")) {
+				// we are in bootstrap mode and must find a previously released
+				// version to kick-off from
+				return installBootstrapRascalVersion();
+			}
+
+			for (Object o : project.getArtifacts()) {
+				Artifact a = (Artifact) o;
+
+				if (a.getArtifactId().equals("org.rascalmpl:rascal")) {
+					File file = a.getFile().getAbsoluteFile();
+
+					if (a.getSelectedVersion().compareTo(getReferenceRascalVersion()) >= 0) {
+						return file.toPath();
+					}
+					else {
+						log.warn("Rascal version in pom.xml dependency is too old for this Rascal maven plugin. " + getReferenceRascalVersion() + " or later expected.");
+						log.warn("Using a newer version org.rascalmpl:rascal:" + bootstrapRascalVersion + "; please add it to the pom.xml");
```

Comment:
why does this differ from `getReferenceRascalVersion()`?

Reply:
the reference is the first version that should support these features. This code resolves the current effective dependency which could be a much newer version.